### PR TITLE
kube-proxy tries to apply latest snapshot

### LINKED ISSUE #20402: Services and endpoints out of sync causing test flake
http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce/10740/

`Services should be able to create a functioning NodePort service` and `Services should be able to change the type and nodeport settings of a service` failed with message:
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:416 Feb 1 00:44:22.421: Could not reach nodePort service through node-ip 130.211.158.115:32285 in 45s
```

cc @thockin @kubernetes/goog-cluster

### LINKED ISSUE #20470: Kubeproxy should always apply latest snapshot
This bug is mostly to document the flow that most likely leads to the latency observed in https://github.com/kubernetes/kubernetes/issues/20402.

1. Watch comes in, sends an update/add/del to proxiers api source: https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/config/api.go#L56
2. ApiSource snapshots all services or endpoints and shoves them in a channel: https://github.com/kubernetes/kubernetes/blob/master/pkg/client/cache/undelta_store.go#L57
3. Merger pulls out of this channel, inserts into shared map, and sends notification down the update channel: https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/config/config.go#L156
4. Proxy notices update, invokes `OnService/EndpointUpdate`: https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/config/config.go#L276

If running OnEndpointsUpdate takes 100ms during which 4 other endpoints come down the watch, we'll perform 1-4 for each. Instead, if the update channel in step 3 is full we can just go back to pulling out new watch snapshots from 1. This will drop intermediate updates.

### PR DESCRIPTION
**Pull request by bprashanth**
Added some logs to kube proxy, and removeda blocking channel operation. 

See https://github.com/kubernetes/kubernetes/issues/20470 for details. Waiting on the pr builder to try it, possible fix for symptoms observed in: https://github.com/kubernetes/kubernetes/issues/20402

closes #20402
closes #20470

### CODE REVIEW

**Comment by bprashanth** [comment]
It's not like we have 100s of services coming up and down, and yet, there's a mysterious backlog looking delay in the logs. So this can't hurt?

**Comment by thockin** [comment]
Is this for commit?  LGTM

**Comment by bprashanth** [comment]
With almost impeccable timing I got a legit e2e failure for nodeports services last time, but there's too much noise from the userspace proxy to zero in on latencies. I do see a lot of "channel full" messages that I'm adding in this pr, so I'm planning to take this a step at a time.

@thockin ptal at last 2 commits (unittest fix and 45s timeout I'd previously missed)

**Comment by bprashanth** [comment]
Second commit has the timeout bump, ref: https://github.com/kubernetes/kubernetes/issues/20567. I can drop the first one if you prefer to keep this just about logs/timeout.

File: pkg/proxy/iptables/proxier.go [hunk 1]
```diff
@@ -347,6 +350,11 @@ func (proxier *Proxier) OnServiceUpdate(allServices []api.Service) {
 
 // OnEndpointsUpdate takes in a slice of updated endpoints.
 func (proxier *Proxier) OnEndpointsUpdate(allEndpoints []api.Endpoints) {
+	start := time.Now()
+	defer func() {
```
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: should we add one of these to the syncProxyRules() ?

File: pkg/proxy/config/config.go [hunk 2]
```diff
@@ -153,7 +153,13 @@ func (s *endpointsStore) Merge(source string, change interface{}) error {
 	s.endpoints[source] = endpoints
 	s.endpointLock.Unlock()
 	if s.updates != nil {
-		s.updates <- struct{}{}
+		// Since we record the snapshot before sending this signal, it's
+		// possible that the consumer ends up performing an extra update.
+		select {
+		case s.updates <- struct{}{}:
+		default:
+			glog.V(4).Infof("Endpoints update channel full.")
```
Comment (thockin) [inline pkg/proxy/config/config.go]: Can you also change this messages to not be "full" (whcih sounds bad) but something like "update flag is already set" ?

**Comment by thockin** [comment]
LGTM - 2 nits then self-LGTM it.

@saad-ali this one should get prio to help diagnose flakes/test-fails

**Comment by thockin** [comment]
Travis is horked.  IT hasn't even started this.

**Comment by saad-ali** [comment]
Manually merging

**Pull request merged**